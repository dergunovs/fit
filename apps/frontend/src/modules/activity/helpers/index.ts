import { IActivity, IExerciseDone, IExerciseStatistics } from 'fitness-tracker-contracts';
import { createTempId, formatDate, formatDuration, subtractDates } from 'mhz-helpers';
import { toast } from 'mhz-ui';

export function getPotentialActivityDuration(
  exercises: IExerciseDone[],
  exerciseStatistics: IExerciseStatistics[],
  averageRestPercent?: number
) {
  if (!averageRestPercent) return '-';

  const totalDuration = exercises.reduce((acc, exercise) => {
    const averageDuration =
      exerciseStatistics.find((choosenExericse) => choosenExericse._id === exercise.exercise?._id)?.averageDuration ||
      0;

    return acc + averageDuration * exercise.repeats;
  }, 0);

  const durationWithRest = Math.round(totalDuration / (1 - averageRestPercent / 100));

  return formatDuration(durationWithRest);
}

export function convertActivityCalendarEvents(activities?: IActivity[]) {
  return activities?.map((activity: IActivity) => {
    return {
      _id: activity._id,
      start: new Date(`${activity.dateCreated}`),
      end: new Date(`${activity.dateUpdated}`),
      title: '+',
      content: activity.exercises,
    };
  });
}

export function generateActivityExercises(exercisesDone: IExerciseDone[]): IExerciseDone[] {
  const activityExercises = exercisesDone.map((exercise) => {
    return {
      _id: createTempId(),
      exercise: {
        _id: exercise.exercise?._id || '',
        title: exercise.exercise?.title || '',
        muscleGroups: exercise.exercise?.muscleGroups || [],
        isWeights: exercise.exercise?.isWeights || false,
        isWeightsRequired: exercise.exercise?.isWeightsRequired || false,
      },
      repeats: exercise.repeats,
      weight: exercise.weight,
    };
  });

  return activityExercises?.length ? [...activityExercises] : [];
}

export function getToFailurePercent(exercises: IExerciseDone[]) {
  const allExercises = exercises.length;
  const toFailureExercises = exercises.filter((exercise) => exercise.isToFailure).length;

  return `${Math.floor((toFailureExercises / allExercises) * 100)}%`;
}

export function getRestPercent(exercises: IExerciseDone[], start?: Date | null | string, end?: Date | null | string) {
  const activityDuration = Number(subtractDates(end, start, true));
  const exercisesDuration = exercises.reduce((acc, current) => acc + (current.duration || 0), 0);

  return `${Math.floor(100 - (exercisesDuration / activityDuration) * 100)}%`;
}

export async function copyActivityToClipboard(
  exercises: IExerciseDone[],
  start?: Date | null | string,
  end?: Date | null | string
) {
  const textHeader = `${formatDate(start, 'ru')}, длительность: ${subtractDates(end, start)}
Подходы: ${exercises.length}, отказы: ${getToFailurePercent(exercises)}, отдых: ${getRestPercent(exercises, start, end)}.

${exercises
  .map((exercise, index) => {
    return `${index + 1}. ${exercise.exercise?.title} x${exercise.repeats} ${exercise.weight ? `${exercise.weight}кг` : ''} ${formatDuration(exercise.duration)} ${exercise.isToFailure ? 'ДО ОТКАЗА' : ''}\n`;
  })
  .join('')}`;

  await navigator.clipboard.writeText(textHeader);

  toast.success('Скопировано в буфер');
}
