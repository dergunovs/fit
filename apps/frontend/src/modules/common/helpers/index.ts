import { toast } from 'mhz-ui';

export async function copyToClipboard(text: string, toastText: string) {
  await navigator.clipboard.writeText(text);

  toast.success(toastText);
}
