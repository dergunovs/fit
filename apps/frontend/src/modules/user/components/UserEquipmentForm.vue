<template>
  <UiFlex gap="16" column>
    <UiFlex>
      <UiFlex column>
        <div>Оборудование</div>

        <UiFlex>
          <UiSelect
            v-model="choosenEquipment"
            :options="equipmentsFiltered"
            :isDisabled="isEditEquipment"
            lang="ru"
            data-test="user-form-equipment-options"
          />
        </UiFlex>
      </UiFlex>

      <UiFlex v-if="choosenEquipment?.isWeights" column>
        <div>Возможный вес</div>

        <UiFlex>
          <UiInput
            v-model="choosenEquipmentWeightInput"
            type="number"
            step="1"
            min="1"
            max="500"
            data-test="user-form-weights"
          />

          <UiButton
            :isDisabled="isAddWeightDisabled"
            layout="secondary"
            isNarrow
            @click="addWeight"
            data-test="user-form-add-weight"
          >
            Добавить
          </UiButton>
        </UiFlex>
      </UiFlex>
    </UiFlex>

    <UiFlex v-if="choosenEquipmentWeights.length" column>
      <div>Добавленные веса</div>

      <UiFlex>
        <UiChip v-for="weight in choosenEquipmentWeights" :key="weight">
          <IconWeight width="16" height="16" /><span>{{ weight }}</span> кг.

          <button
            type="button"
            @click="deleteWeight(weight)"
            :class="$style.delete"
            data-test="user-form-delete-weight"
          >
            ×
          </button>
        </UiChip>
      </UiFlex>
    </UiFlex>

    <UiFlex>
      <UiButton v-if="isEditEquipment" @click="saveEquipment" data-test="user-form-save-equipment">
        Сохранить оборудование
      </UiButton>

      <UiButton v-if="isEditEquipment" @click="resetEquipment" layout="secondary" data-test="user-form-reset-equipment">
        Отменить редактирование
      </UiButton>

      <UiButton v-else :isDisabled="isAddEquipmentDisabled" @click="addEquipment" data-test="user-form-add-equipment">
        Добавить оборудование
      </UiButton>
    </UiFlex>

    <UiFlex v-if="props.modelValue?.length" column>
      <div>Добавленное оборудование</div>

      <UiFlex wrap>
        <UiChip
          v-for="equipment in props.modelValue"
          :key="`${equipment.equipment?._id}-${equipment.weights?.join()}`"
          data-test="user-form-equipment"
        >
          <span data-test="user-form-equipment-title">{{ equipment.equipment?.title }}</span>

          <span v-for="weight in equipment.weights" :key="weight" data-test="user-form-equipment-weights">
            <span data-test="user-form-equipment-weight">{{ weight }}</span> кг.
          </span>

          <button
            v-if="equipment.equipment?.isWeights"
            type="button"
            @click="editEquipment(equipment)"
            :class="$style.edit"
            data-test="user-form-equipment-edit"
          >
            <IconEdit width="20" height="20" />
          </button>

          <button
            v-if="equipment.equipment?.title"
            type="button"
            @click="deleteEquipment(equipment.equipment.title)"
            :class="$style.delete"
            data-test="user-form-equipment-delete"
          >
            ×
          </button>
        </UiChip>
      </UiFlex>
    </UiFlex>
  </UiFlex>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';
import { IEquipment, IUserEquipment } from 'fitness-tracker-contracts';
import { UiButton, UiFlex, UiInput, UiSelect, UiChip } from 'mhz-ui';

import IconWeight from '@/common/icons/weight.svg';
import IconEdit from '@/common/icons/edit.svg';

interface IProps {
  equipments: IEquipment[];
  modelValue?: IUserEquipment[];
}

const props = defineProps<IProps>();
const emit = defineEmits<{ 'update:modelValue': [value?: IUserEquipment[]] }>();

const choosenEquipment = ref<IEquipment>();
const choosenEquipmentWeightInput = ref<number>(0);
const choosenEquipmentWeights = ref<number[]>([]);

const isEditEquipment = ref(false);

const equipmentsFiltered = computed(() =>
  props.equipments.filter(
    (equipment) =>
      !props.modelValue?.some((equipmentToFilter) => equipment.title === equipmentToFilter.equipment?.title)
  )
);

const isAddWeightDisabled = computed(
  () =>
    choosenEquipmentWeightInput.value === 0 ||
    choosenEquipmentWeights.value.includes(Number(choosenEquipmentWeightInput.value))
);

const isAddEquipmentDisabled = computed(
  () =>
    !choosenEquipment.value ||
    (choosenEquipment.value.isWeights && !choosenEquipmentWeights.value.length) ||
    props.modelValue?.some((equipment) => equipment.equipment?.title === choosenEquipment.value?.title)
);

function addWeight() {
  if (isAddWeightDisabled.value) return;

  choosenEquipmentWeights.value = [...choosenEquipmentWeights.value, Number(choosenEquipmentWeightInput.value)].sort();
}

function deleteWeight(weightToDelete: number) {
  choosenEquipmentWeights.value = choosenEquipmentWeights.value.filter((weight) => weight !== weightToDelete);
}

function addEquipment() {
  if (isAddEquipmentDisabled.value) return;

  const updatedEquipments = props.modelValue?.length
    ? [...props.modelValue, { equipment: choosenEquipment.value, weights: choosenEquipmentWeights.value }]
    : [{ equipment: choosenEquipment.value, weights: choosenEquipmentWeights.value }];

  emit('update:modelValue', updatedEquipments);

  resetEquipment();
}

function editEquipment(equipment: IUserEquipment) {
  isEditEquipment.value = true;

  choosenEquipment.value = equipment.equipment;
  choosenEquipmentWeightInput.value = 0;
  choosenEquipmentWeights.value = equipment.weights?.length ? [...equipment.weights] : [];
}

function saveEquipment() {
  const updatedEquipments = props.modelValue?.map((equipment) => {
    if (equipment.equipment?.title === choosenEquipment.value?.title) {
      return { equipment: choosenEquipment.value, weights: choosenEquipmentWeights.value };
    } else return equipment;
  });

  emit('update:modelValue', updatedEquipments);

  resetEquipment();
}

function deleteEquipment(title: string) {
  const updatedEquipments = props.modelValue?.filter((equipment) => equipment.equipment?.title !== title);

  emit('update:modelValue', updatedEquipments);
}

function resetEquipment() {
  choosenEquipment.value = undefined;
  choosenEquipmentWeightInput.value = 0;
  choosenEquipmentWeights.value = [];

  isEditEquipment.value = false;
}
</script>

<style module lang="scss">
.edit {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 2px;
  cursor: pointer;
  background: none;
  border: none;
}

.delete {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 20px;
  height: 20px;
  font-size: 0.75rem;
  font-weight: 700;
  color: var(--color-white);
  cursor: pointer;
  background-color: var(--color-error);
  border: none;
  border-radius: 50%;
}
</style>
