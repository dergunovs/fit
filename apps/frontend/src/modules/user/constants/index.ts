export const URL_USER = '/users';
export const URL_USER_CREATE = `${URL_USER}/create`;
export const URL_USER_EDIT = `${URL_USER}/edit`;
export const URL_USER_PROFILE = `${URL_USER}/profile`;

export const USER_GOALS_OPTIONS = {
  activities: [1, 2, 3, 4, 5],
  sets: [12, 16, 20, 24, 30],
  repeats: [8, 10, 12, 15, 20],
  duration: [20, 30, 40, 50, 60],
};
