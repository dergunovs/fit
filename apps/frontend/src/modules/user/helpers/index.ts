import { IEquipment, IUserEquipment } from 'fitness-tracker-contracts';

export function excludeChoosenUserEquipment(equipments: IEquipment[], userEquipments?: IUserEquipment[]) {
  return equipments.filter(
    (equipment) => !userEquipments?.some((equipmentToFilter) => equipment._id === equipmentToFilter.equipment?._id)
  );
}
