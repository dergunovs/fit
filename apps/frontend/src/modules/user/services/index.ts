import { ComputedRef, Ref } from 'vue';
import {
  API_USER,
  API_USER_PASSWORD,
  TGetUsersDTO,
  TGetUsersQueryDTO,
  TGetUserDTO,
  TPostUserDTO,
  TPostUserDataDTO,
  TUpdateUserDTO,
  TUpdateUserDataDTO,
  TDeleteUserDTO,
  TUpdateUserPasswordDTO,
  TUpdateUserPasswordDataDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const userService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_USER, page],
      queryFn: async () => {
        const params: TGetUsersQueryDTO = { page: page.value };

        const { data } = await api.get<TGetUsersDTO>(API_USER, { params });

        return data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined> | Ref<string | undefined>) => {
    return useQuery({
      queryKey: [API_USER, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetUserDTO>(`${API_USER}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (formData: TPostUserDataDTO) => {
        const { data } = await api.post<TPostUserDTO>(API_USER, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (formData: TUpdateUserDataDTO) => {
        const { data } = await api.patch<TUpdateUserDTO>(`${API_USER}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  updatePassword: (options: object, id?: string) => {
    return useMutation({
      mutationKey: [API_USER_PASSWORD, id],
      mutationFn: async (password: TUpdateUserPasswordDataDTO) => {
        if (!id) return null;

        const { data } = await api.patch<TUpdateUserPasswordDTO>(`${API_USER_PASSWORD}/${id}`, password);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteUserDTO>(`${API_USER}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
