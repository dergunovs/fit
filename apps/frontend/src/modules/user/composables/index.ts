import { computed, Ref } from 'vue';

import { useTI18n } from '@/common/composables';

export function useUserFormTabs(isEdit: boolean, isAdmin: Ref<boolean>) {
  const { t } = useTI18n();

  const generalTab = computed(() => {
    return { value: 'general', title: t('user.general') };
  });

  const exercisesTab = computed(() => {
    return { value: 'exercises', title: t('exercise.many') };
  });

  const otherTabs = computed(() => [
    { value: 'goals', title: t('goals') },
    { value: 'weight', title: t('weight') },
    { value: 'templates', title: t('template.many') },
  ]);

  const TABS = computed(() => {
    if (!isEdit) return [generalTab.value];
    if (isAdmin.value) return [generalTab.value, ...otherTabs.value];

    return [generalTab.value, ...otherTabs.value, exercisesTab.value];
  });

  return { TABS };
}
