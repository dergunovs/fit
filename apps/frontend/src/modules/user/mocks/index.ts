import { vi } from 'vitest';
import {
  TDeleteUserDTO,
  TGetUsersDTO,
  TPostUserDataDTO,
  TPostUserDTO,
  TUpdateUserDataDTO,
  TUpdateUserDTO,
} from 'fitness-tracker-contracts';

import { userService } from '@/user/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { USER_FIXTURE, USERS_FIXTURE } from '@/user/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetUser = vi.spyOn(userService, 'getOne').mockReturnValue(mockQueryReply(USER_FIXTURE));

const mockGetUsersData: TGetUsersDTO = { data: USERS_FIXTURE, total: USERS_FIXTURE.length };

const spyGetUsers = vi.spyOn(userService, 'getMany').mockImplementation(() => mockQueryReply(mockGetUsersData));

const spyCreateUser = vi.fn();
const spyUpdateUser = vi.fn();
const spyDeleteUser = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(userService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostUserDTO, TPostUserDataDTO>(spyCreateUser);
});

vi.spyOn(userService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateUserDTO, TUpdateUserDataDTO>(spyUpdateUser);
});

vi.spyOn(userService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteUserDTO, string>(spyDeleteUser);
});

export { spyGetUser, spyGetUsers, spyCreateUser, spyUpdateUser, spyDeleteUser, mockOnSuccess, mockGetUsersData };
