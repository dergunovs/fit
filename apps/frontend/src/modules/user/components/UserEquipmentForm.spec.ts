import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { IEquipment } from 'fitness-tracker-contracts';
import { dataTest } from 'mhz-helpers';

import UserEquipmentForm from './UserEquipmentForm.vue';

import { wrapperFactory } from '@/common/test';
import { EQUIPMENTS_FIXTURE } from '@/equipment/fixtures';
import { USER_FIXTURE } from '@/user/fixtures';
import { excludeChoosenUserEquipment } from '@/user/helpers';

const weightsCount = USER_FIXTURE.equipments?.[0].weights?.length || 0;

const options = dataTest('user-equipment-options');
const edit = dataTest('user-equipment-edit');
const weight = dataTest('user-weight');
const addWeight = dataTest('user-add-weight');
const addedWeights = dataTest('user-added-weights');
const addedWeight = dataTest('user-added-weight');
const deleteWeight = dataTest('user-delete-weight');
const equipment = dataTest('user-equipment');
const equipmentTitle = dataTest('user-equipment-title');
const equipmentWeights = dataTest('user-equipment-weights');
const equipmentWeight = dataTest('user-equipment-weight');
const save = dataTest('user-save-equipment');
const reset = dataTest('user-reset-equipment');
const addEquipment = dataTest('user-add-equipment');
const deleteEquipment = dataTest('user-delete-equipment');

let wrapper: VueWrapper<InstanceType<typeof UserEquipmentForm>>;

beforeEach(() => {
  wrapper = wrapperFactory(UserEquipmentForm, { equipments: EQUIPMENTS_FIXTURE, modelValue: USER_FIXTURE.equipments });
});

enableAutoUnmount(afterEach);

describe('UserEquipmentForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(UserEquipmentForm)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('sets filtered equipment options', async () => {
    expect(wrapper.find(options).attributes('options')).toStrictEqual(
      excludeChoosenUserEquipment(EQUIPMENTS_FIXTURE, USER_FIXTURE.equipments).join('')
    );
  });

  it('disables equipment select when edit', async () => {
    expect(wrapper.find(options).attributes('isdisabled')).toStrictEqual('false');
    expect(wrapper.find(options).attributes('modelvalue')).toStrictEqual(undefined);
    expect((wrapper.vm as unknown as { choosenEquipment: IEquipment }).choosenEquipment).toStrictEqual(undefined);

    await wrapper.find(edit).trigger('click');

    expect(wrapper.find(options).attributes('isdisabled')).toStrictEqual('true');
    expect(wrapper.find(options).attributes('modelvalue')).not.toStrictEqual(undefined);
    expect((wrapper.vm as unknown as { choosenEquipment: IEquipment }).choosenEquipment).toStrictEqual(
      USER_FIXTURE.equipments?.[0].equipment
    );

    await wrapper.find(reset).trigger('click');

    expect(wrapper.find(options).attributes('isdisabled')).toStrictEqual('false');
    expect(wrapper.find(options).attributes('modelvalue')).toStrictEqual(undefined);
  });

  it('shows weight input when equipment can have a weight', async () => {
    expect(wrapper.find(weight).exists()).toBe(false);

    await wrapper.find(edit).trigger('click');

    expect(wrapper.find(weight).exists()).toBe(true);
  });

  it('adds new weight to equipment with validation', async () => {
    const OLD_WEIGHT = USER_FIXTURE.equipments?.[0].weights?.[0];
    const NEW_WEIGHT = 7;

    await wrapper.find(edit).trigger('click');

    expect(wrapper.find(addWeight).attributes('isdisabled')).toStrictEqual('true');

    await wrapper.findComponent(weight).setValue(OLD_WEIGHT);

    expect(wrapper.find(addWeight).attributes('isdisabled')).toStrictEqual('true');

    await wrapper.findComponent(weight).setValue(NEW_WEIGHT);

    expect(wrapper.find(addWeight).attributes('isdisabled')).toStrictEqual('false');

    expect(wrapper.findAll(addedWeights).length).toBe(weightsCount);

    await wrapper.find(addWeight).trigger('click');

    expect(wrapper.findAll(addedWeights).length).toBe(weightsCount + 1);
    expect(wrapper.findAll(addedWeight)[weightsCount].text()).toBe(NEW_WEIGHT.toString());

    await wrapper.find(save).trigger('click');

    if (!USER_FIXTURE.equipments || !USER_FIXTURE.equipments[0].weights) return;

    const updatedWeights = [...USER_FIXTURE.equipments[0].weights, NEW_WEIGHT];

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual([
      [{ equipment: USER_FIXTURE.equipments[0].equipment, weights: updatedWeights }, USER_FIXTURE.equipments[1]],
    ]);
  });

  it('deletes weight', async () => {
    await wrapper.find(edit).trigger('click');

    expect(wrapper.findAll(addedWeights).length).toBe(weightsCount);

    await wrapper.findAll(deleteWeight)[weightsCount - 1].trigger('click');

    expect(wrapper.findAll(addedWeights).length).toBe(weightsCount - 1);

    await wrapper.find(save).trigger('click');

    if (!USER_FIXTURE.equipments || !USER_FIXTURE.equipments[0].weights) return;

    const updatedWeights = USER_FIXTURE.equipments[0].weights.slice(0, -1);

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual([
      [{ equipment: USER_FIXTURE.equipments[0].equipment, weights: updatedWeights }, USER_FIXTURE.equipments[1]],
    ]);
  });

  it('deletes and adds equipment', async () => {
    if (!USER_FIXTURE.equipments) return;

    await wrapper.findAll(deleteEquipment)[USER_FIXTURE.equipments?.length - 1].trigger('click');

    const updatedEquipments = USER_FIXTURE.equipments.slice(0, -1);

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual([updatedEquipments]);

    (wrapper.vm as unknown as { choosenEquipment?: IEquipment }).choosenEquipment =
      USER_FIXTURE.equipments[1].equipment;

    await wrapper.findComponent(options).setValue(USER_FIXTURE.equipments[1].equipment);
    await wrapper.setProps({ modelValue: updatedEquipments });

    await wrapper.find(addEquipment).trigger('click');

    expect(wrapper.emitted('update:modelValue')).toHaveLength(2);
    expect(wrapper.emitted()['update:modelValue'][1]).toStrictEqual([USER_FIXTURE.equipments]);
  });

  it('hides equipment when modelValue is empty', async () => {
    expect(wrapper.findAll(equipment).length).toBe(USER_FIXTURE.equipments?.length);
    expect(wrapper.find(equipmentTitle).text()).toBe(USER_FIXTURE.equipments?.[0].equipment?.title);
    expect(wrapper.findAll(equipmentWeights).length).toBe(USER_FIXTURE.equipments?.[0].weights?.length);
    expect(wrapper.find(equipmentWeight).text()).toBe(USER_FIXTURE.equipments?.[0].weights?.[0].toString());

    await wrapper.setProps({ modelValue: [] });

    expect(wrapper.find(equipment).exists()).toBe(false);
    expect(wrapper.find(equipmentTitle).exists()).toBe(false);
    expect(wrapper.find(equipmentWeights).exists()).toBe(false);
    expect(wrapper.find(equipmentWeight).exists()).toBe(false);
  });

  it('disables add weight button when weight is invalid (less than 1)', async () => {
    await wrapper.find(edit).trigger('click');

    await wrapper.findComponent(weight).setValue(-1);

    expect(wrapper.find(addWeight).attributes('isdisabled')).toStrictEqual('true');
  });

  it('disables add weight button when weight is invalid (greater than 500)', async () => {
    await wrapper.find(edit).trigger('click');

    await wrapper.findComponent(weight).setValue(501);

    expect(wrapper.find(addWeight).attributes('isdisabled')).toStrictEqual('true');
  });

  it('disables add weight button when weight is not a number', async () => {
    await wrapper.find(edit).trigger('click');

    await wrapper.findComponent(weight).setValue('abc');

    expect(wrapper.find(addWeight).attributes('isdisabled')).toStrictEqual('true');
  });
});
