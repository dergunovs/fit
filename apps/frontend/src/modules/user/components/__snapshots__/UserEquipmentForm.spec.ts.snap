// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserEquipmentForm > matches snapshot 1`] = `
"<div gap="16" column="">
  <div>
    <div column="">
      <div>Оборудование</div>
      <div><input type="text" options="" isdisabled="false" lang="ru" islocalefield="false" data-test="user-equipment-options"></div>
    </div>
    <!--v-if-->
  </div>
  <!--v-if-->
  <!--v-if-->
  <div>
    <!--v-if--><button isdisabled="true" data-test="user-add-equipment">Добавить оборудование</button>
  </div>
  <div column="">
    <div>Оборудование добавлено</div>
    <div wrap="">
      <div isedit="true" isdelete="" data-test="user-equipment"><span data-test="user-equipment-title">Гантели</span><span data-test="user-equipment-weights"><span data-test="user-equipment-weight">1</span> кг</span><span data-test="user-equipment-weights"><span data-test="user-equipment-weight">2</span> кг</span><span data-test="user-equipment-weights"><span data-test="user-equipment-weight">3</span> кг</span></div>
      <div isedit="false" isdelete="" data-test="user-equipment"><span data-test="user-equipment-title">Скамья</span></div>
    </div>
  </div>
</div>"
`;
